util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+	DbgPrint("getNodesContainer: schema %s, next path %v, value %v", schema.Name, path.GetElem()[0], ValueStr(root))
+
+	rv := reflect.ValueOf(root)
+	if !IsValueStructPtr(rv) {
+		return nil, nil, fmt.Errorf("getNodesContainer: root has type %T, expect struct ptr", root)
+	}
+
+	v := rv.Elem()
+
+	for i := 0; i < v.NumField(); i++ {
+		f := v.Field(i)
+		ft := v.Type().Field(i)
+		cschema, err := FieldSchema(schema, ft)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error for schema for type %T, field name %s: %s", root, ft.Name, err)
+		}
+		if cschema == nil {
+			return nil, nil, fmt.Errorf("could not find schema for type %T, field name %s", root, ft.Name)
+		}
+		cschema, err = ResolveLeafRef(cschema)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error for schema for type %T, field name %s: %s", root, ft.Name, err)
+		}
+
+		DbgPrint("check field name %s", cschema.Name)
+		ps, err := SchemaPaths(ft)
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, p := range ps {
+			if pathMatchesPrefix(path, p) {
+				// don't trim whole prefix  for keyed list since name and key
+				// are a in the same element.
+				to := len(p)
+				if IsTypeMap(ft.Type) {
+					to--
+				}
+				return getNodesInternal(cschema, f.Interface(), trimGNMIPathPrefix(path, p[0:to]))
+			}
+		}
+	}
+
+	return nil, nil, DbgErr(fmt.Errorf("could not find path in tree beyond schema node %s, (type %T), remaining path %v", schema.Name, root, path))
+}
+
+// getNodesList traverses the list root, which must be a map of struct
+// type and matches each map key against the first path element in path. If the
+// key matches, it recurses into that field with the remaining path.
+// If empty key is specified, all list elements match.
+func getNodesList(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	DbgPrint("getNodesList: schema %s, next path %v, value %v", schema.Name, path.GetElem()[0], ValueStr(root))
+
+	rv := reflect.ValueOf(root)
+	if schema.Key == "" {
+		return nil, nil, fmt.Errorf("getNodesList: path %v cannot traverse unkeyed list type %T", path, root)
+	}
+	if !IsValueMap(rv) {
+		// Only keyed lists can be traversed with a path.
+		return nil, nil, fmt.Errorf("getNodesList: root has type %T, expect map", root)
+	}
+	emptyKey := false
+	if len(path.GetElem()[0].GetKey()) == 0 {
+		DbgPrint("path %v at %T points to list with empty wildcard key", path, root)
+		emptyKey = true
+	}
+
+	listElementType := rv.Type().Elem().Elem()
+	listKeyType := rv.Type().Key()
+
+	var matchNodes []interface{}
+	var matchSchemas []*yang.Entry
+
+	// Iterate through all the map keys to see if any match the path.
+	for _, k := range rv.MapKeys() {
+		ev := rv.MapIndex(k)
+		DbgPrint("checking key %v, value %v", k.Interface(), ValueStr(ev.Interface()))
+		match := true
+		if !emptyKey { // empty key matches everything.
+			if !IsValueStruct(k) {
```

Comment:
In the generated code there is a function on each `GoStruct` which represents a list member called [`ΛListKeyMap`](https://github.com/openconfig/ygot/blob/master/ygen/gogen.go#L511) -- this returns `map[string]interface{}` of the keys that are specified within the struct (keyed by schema name). It seems like you might be able to simplify here by calling that function and then just checking map equality (or partial match).

The `Sprintf` below is probably still required however.

Reply:
Comparing the code required to reflect into ΛListKeyMap it doesn't look much better than what's here and it probably makes the logic even harder to understand for a casual reader (they would have to know what ΛListKeyMap contains etc.)

Reply:
Sure - your choice :-)